File: ratelimiter/src/main/java/com/evalvis/ratelimiter/web/HopByHopHeaders.java
package com.evalvis.ratelimiter.web;

import java.util.Locale;
import java.util.Set;
import org.springframework.http.HttpHeaders;

public final class HopByHopHeaders {

	private static final Set<String> REQUEST_SKIP = Set.of(
		"connection",
		"keep-alive",
		"proxy-connection",
		"transfer-encoding",
		"upgrade",
		"host",
		"content-length"
	);

	private static final Set<String> RESPONSE_SKIP = Set.of(
		"connection",
		"keep-alive",
		"proxy-connection",
		"transfer-encoding",
		"upgrade"
	);

	private HopByHopHeaders() {
	}

	public static boolean skipRequestHeader(String name) {
		return REQUEST_SKIP.contains(name.toLowerCase(Locale.ROOT));
	}

	public static HttpHeaders filterResponse(HttpHeaders source) {
		HttpHeaders out = new HttpHeaders();
		source.forEach((name, values) -> {
			if (!RESPONSE_SKIP.contains(name.toLowerCase(Locale.ROOT))) {
				out.put(name, values);
			}
		});
		return out;
	}

}

File: loadbalancer/src/main/kotlin/com/evalvis/loadbalancer/web/HopByHopHeaders.kt
package com.evalvis.loadbalancer.web

import org.springframework.http.HttpHeaders
import java.util.Locale

object HopByHopHeaders {

	private val requestSkip = setOf(
		"connection",
		"keep-alive",
		"proxy-connection",
		"transfer-encoding",
		"upgrade",
		"host",
		"content-length",
	)

	private val responseSkip = setOf(
		"connection",
		"keep-alive",
		"proxy-connection",
		"transfer-encoding",
		"upgrade",
	)

	fun skipRequestHeader(name: String): Boolean {
		return requestSkip.contains(name.lowercase(Locale.ROOT))
	}

	fun filterResponse(source: HttpHeaders): HttpHeaders {
		val out = HttpHeaders()
		source.forEach { name, values ->
			if (!responseSkip.contains(name.lowercase(Locale.ROOT))) {
				out.addAll(name, values)
			}
		}
		return out
	}
}
